src/assocPath.js

```diff
@@ -41,7 +43,8 @@ module.exports = _curry3(function assocPath(path, val, obj) {
   }
   if (_isInteger(idx) && _isArray(obj)) {
     var arr = [].concat(obj);
-    arr[idx] = val;
+    var arrIdx = idx < 0 ? arr.length + idx : idx;
```

Comment:
I'm not sure. I feel like the logic of associating a value to an index of a flat array or a property given an flat object is sufficiently complex to warrant it's own function (`assoc`). `assocPath` could use that function when `path.length === 1`. If we put all the logic in `assocPath` then I think we are going to end up with quite a complex function. But perhaps I'm missing a simple solution. 

I'll push up a solution I think is good and you can let me know what you think.